File: implementations/fastapi/examples/src/account/application/command/apply_daily_interest_handler.py
from __future__ import annotations

from datetime import date
from decimal import Decimal

from ...domain.repository import AccountRepository

PAGE_SIZE = 200


class ApplyDailyInterestHandler:
    """The Command Service for the regular interest-payment batch. Since this is a
    system-driven use case triggered once a day by the Task Queue
    (account.interest.apply), there is no Command dataclass representing a user request
    (the convention for other Handlers) — because there is no HTTP requester.

    Iterates every ACTIVE account page by page, calling Account.apply_interest(). Calls
    save_account() for every account regardless of whether interest was actually paid or
    its amount — the Handler doesn't need to distinguish whether apply_interest() updated
    `last_interest_paid_at` internally (an account already processed today doesn't even
    touch the field, so re-saving is effectively a no-op UPDATE).
    """

    def __init__(self, repo: AccountRepository) -> None:
        self._repo = repo

    async def execute(self, daily_rate: Decimal, today: date | None = None) -> int:
        today = today or date.today()
        applied_count = 0
        page = 0

        while True:
            accounts, total = await self._repo.find_accounts(page=page, take=PAGE_SIZE, status=["ACTIVE"])
            if not accounts:
                break

            for account in accounts:
                transaction = account.apply_interest(daily_rate, today)
                await self._repo.save_account(account)
                if transaction is not None:
                    applied_count += 1

            if (page + 1) * PAGE_SIZE >= total:
                break
            page += 1

        return applied_count

File: implementations/go/examples/internal/application/command/apply_daily_interest_handler.go
package command

import (
	"context"
	"fmt"
	"time"

	"github.com/example/account-service/internal/domain/account"
)

// ApplyDailyInterestCommand is the input InterestTaskController (interface/task/)
// builds by deserializing the Task Queue message payload. Date is in
// "2006-01-02" format, and InterestScheduler.EnqueueDailyInterest sends the
// date as of enqueue time as-is in the payload — so even if the Consumer
// processes this message later (past midnight), "which date's interest this
// is" stays fixed.
type ApplyDailyInterestCommand struct {
	Date string
}

// ApplyDailyInterestHandler is a system-triggered use case driven by the Task
// Queue — since it is not a user command, it does not take authorization
// info such as RequesterID. It pages through all ACTIVE accounts and calls
// Account.ApplyInterest (Level 1 idempotent) on each.
type ApplyDailyInterestHandler struct {
	repo account.Repository
	rate float64
}

func NewApplyDailyInterestHandler(repo account.Repository, rate float64) *ApplyDailyInterestHandler {
	return &ApplyDailyInterestHandler{repo: repo, rate: rate}
}

// interestBatchSize is the number of accounts fetched per page. It is set
// smaller than the Take:1000 idiom used by existing batch-style handlers
// (e.g. suspend_cards_by_account_handler.go) so that per-page query/processing
// time doesn't grow excessive in environments with a large number of accounts.
const interestBatchSize = 500

// Handle saves and returns immediately — publishing to/consuming from SQS via
// the Outbox is solely the responsibility of the independently, periodically
// running outbox.Poller/outbox.Consumer (no synchronous draining,
// domain-events.md). The InterestPaid Domain Event raised by
// Account.ApplyInterest rides this same path — the Task Queue (this batch
// itself) and the Domain Event (the resulting fact of interest having been
// paid) are two distinct mechanisms used together within a single use case.
func (h *ApplyDailyInterestHandler) Handle(ctx context.Context, cmd ApplyDailyInterestCommand) error {
	today, err := time.Parse("2006-01-02", cmd.Date)
	if err != nil {
		return fmt.Errorf("apply daily interest: %w", account.ErrInvalidInterestDate)
	}

	for page := 0; ; page++ {
		accounts, total, err := h.repo.FindAccounts(ctx, account.FindQuery{
			Status: []account.Status{account.StatusActive},
			Take:   interestBatchSize,
			Page:   page,
		})
		if err != nil {
			return fmt.Errorf("apply daily interest: find accounts: %w", err)
		}

		for _, a := range accounts {
			_, applied, err := a.ApplyInterest(h.rate, today)
			if err != nil {
				return fmt.Errorf("apply daily interest: %w", err)
			}
			if !applied {
				continue // Skip: interest already paid today (Level 1 idempotent) or computed interest is 0.
			}
			if err := h.repo.SaveAccount(ctx, a); err != nil {
				return fmt.Errorf("apply daily interest: save account: %w", err)
			}
		}

		if len(accounts) == 0 || (page+1)*interestBatchSize >= total {
			break
		}
	}
	return nil
}
